packages/phone-number-privacy/signer/src/signer/sign.abstract.ts

```diff
@@ -0,0 +1,62 @@
+import {
+  DomainRestrictedSignatureRequest,
+  KEY_VERSION_HEADER,
+  SignMessageRequest,
+  WarningMessage,
+} from '@celo/phone-number-privacy-common'
+import Logger from 'bunyan'
+import { Request } from 'express'
+import { computeBlindedSignature } from '../bls/bls-cryptography-client'
+import { getKeyProvider } from '../key-management/key-provider'
+import { Key } from '../key-management/key-provider-base'
+import { IAction, Session } from './action.interface'
+import { IOAbstract } from './io.abstract'
+
+declare type OdisSignatureRequest = SignMessageRequest | DomainRestrictedSignatureRequest
+
+export abstract class SignAbstract<R extends OdisSignatureRequest> implements IAction<R> {
+  abstract readonly io: IOAbstract<R>
+
+  public abstract perform(session: Session<R>): Promise<void>
+
+  protected async sign(
+    blindedMessage: string,
+    defaultKey: Key,
+    session: Session<R>
+  ): Promise<{ signature: string; key: Key }> {
+    let keyVersion = this.getRequestKeyVersion(session.request, session.logger)
+    if (keyVersion ?? false) {
+      keyVersion = defaultKey.version
+    }
+    const key: Key = { name: defaultKey.name, version: keyVersion! }
+    let privateKey: string
+    try {
+      privateKey = await getKeyProvider().getPrivateKeyOrFetchFromStore(key)
+    } catch (err) {
+      if (key.version !== defaultKey.version) {
+        // @victor Is this the correct behavior?
```

Comment:
If the requested key version is not supported, we should abort the request. The client (e.g. the combiner) asked for a particular version, and it will not help them to respond with a different version.